src/coreclr/vm/eventtrace.cpp

```diff
@@ -2403,10 +2410,25 @@ VOID EtwCallbackCommon(
         GCHeapUtilities::RecordEventStateChange(bIsPublicTraceHandle, keywords, level);
     }
 
-    // Special check for the runtime provider's ManagedHeapCollectKeyword.  Profilers
-    // flick this to force a full GC.
-    if (g_fEEStarted && !g_fEEShutDown && bIsPublicTraceHandle &&
-        ((MatchAnyKeyword & CLR_MANAGEDHEAPCOLLECT_KEYWORD) != 0))
+    // Special check for a profiler requested GC.
+    // A full GC will be forced if:
+    // 1. The runtime has started and is not shutting down.
+    // 2. The public provider is requesting GC.
+    // 3. The provider's ManagedHeapCollectKeyword is enabled.
+    // 4. If it is an ETW provider, the control code is to enable or capture the state of the provider.
+    // 5. If it is an EventPipe provider, the session is not being disabled.
+    bool bValidGCRequest =
+        g_fEEStarted && !g_fEEShutDown &&
+        bIsPublicTraceHandle &&
+        ((MatchAnyKeyword & CLR_MANAGEDHEAPCOLLECT_KEYWORD) != 0) &&
+#if !defined(HOST_UNIX)
+        ((ControlCode == EVENT_CONTROL_CODE_ENABLE_PROVIDER) ||
```

Comment:
When you have multiple sessions running via ETW and disable only one of them, does it report ENABLE_PROVIDER or DISABLE_PROVIDER in the callback? I'm trying to figure out if there are still going to be some cases where it is impossible for us to disambiguate because of limitations in the ETW provided data. I'm also curious if we need to maintain both a ControlCode and a SessionChange parameter when they seem to have overlapping info.

Reply:
After enabling multiple ETW trace sessions, disabling one of them will hit here for the DotNETRuntime provider 2x 
with MatchAnyFilter (0x8000000000000008 and 0x0000000000000008) both with ControlCode 0x00000001.

I also noticed that trying to close the second session, it doesn't actually hit this path as it seemed to have failed. 
![Screenshot 2024-11-08 150212](https://github.com/user-attachments/assets/7b2cac9f-1087-410c-84cf-d68e2d4cc53b)

Maybe because both ETW sessions I used the same command and so it already triggered rundown?

Reply:
OK, so it sounds like with ETW and multiple sessions using the ENABLE_PROVIDER control code will just be unavoidably ambiguous and the runtime would continue to trigger multiple GCs in that case. ETW tools users would need to use the CAPTURE_STATE control code instead if they wanted to avoid the duplicates. For EventPipe at least the issue is resolved.

I was pondering if there was some better way to factor ControlCode, SessionChange, and the isEventPipeCallback bit but I concluded the way you have it seems good. A couple improvements I would suggest:
- Lets define the same values for EVENT_CONTROL_CODE_* on HOST_UNIX too so that we don't have to ifdef the checks. We could do that in eventtracebase.h. Above EventPipe was asserting ControlCode == 0 || ControlCode == 1 and ideally that would be use the named defines too rather than hardcoding magic constants.
- I would add some explicit documentation above the EtwCommonCallback describing that even when a session is disabled the control code will still be ENABLE_PROVIDER if some other session remains active. I think mis-interpreting that parameter has caused confusion and bugs on multiple occasions.

Reply:
> I also noticed that trying to close the second session, it doesn't actually hit this path as it seemed to have failed.

Not sure what would cause that. It certainly might be a bug, but if the behavior is the same before and after this change I'm fine assuming it is orthogonal to the current issue.

Reply:
Sorry, just to clarify my previous statement, disabling a session after having multiple sessions enabled will hit the callback and evaluate `bValidGCRequest` but will NOT trigger a ForceGC, as the MatchAnyKeywords will not have the bitmask enabled.

Going through the scenario of enabling multiple ETW sessions with the provider `Microsoft-Windows-DotNETRuntime:800000`, I'm observing the following sequence:

### Enabling First PerfView ETW Session with keyword

| ControlCode | MatchAnyKeyword | ProviderIndex | Level | ForceGC |
|--------|--------|--------|--------|--------|
| 1 | 0x8000 0000 0000 0000 | DotNETRuntime | 0xff | NO |
| 1 | 0x8000 0000 0080 0000 | DotNETRuntime | 0xff | YES |

### Enabling Second PerfView ETW Session with keyword

| ControlCode | MatchAnyKeyword | ProviderIndex | Level | ForceGC |
|--------|--------|--------|--------|--------|
| 1 | 0x0000 0000 0080 0000 | DotNETRuntime | 0x05 | YES |
| 1 | 0x8000 0000 0080 0000 | DotNETRuntime | 0xff | YES |
| 1 | 0x0000 0000 0000 0000 | DotNETRuntime | 0xff | NO |
| 1 | 0x8000 0000 0080 0000 | DotNETRuntime | 0xff | YES |

### Disabling one of the PerfView ETW Sessions (Stopping with 's')

| ControlCode | MatchAnyKeyword | ProviderIndex | Level | ForceGC |
|--------|--------|--------|--------|--------|
| 1 | 0x8000 0000 0000 0000 | DotNETRuntime | 0xff | NO |
| 1 | 0x8000 0000 0000 0000 | DotNETRuntimeRundown | 0xff | NO |
| 1 | 0x0000 0000 0000 0000 | DotNETRuntime | 0x05 | NO |
| 0 | 0x0000 0000 0000 0000 | DotNETRuntimeRundown | 0x00 | NO |

For what its worth, there was one instance where I observed a ControlCode 0 for DotNETRuntime when disabling the ETW session, and that had Level 0x00, but after many attempts, I haven't been able to reproduce it.

Reply:
Added the keywords to HOST_UNIX and added the note to EtwCallbackCommon